Content.Client/Chat/ChatBox.cs

```diff
@@ -134,27 +508,227 @@ private void InputKeyBindDown(GUIBoundKeyEventArgs args)
+
+            if (Parent == null || _lobbyMode)
             {
                 return;
             }
 
-            var formatted = new FormattedMessage(3);
-            formatted.PushColor(color);
-            formatted.AddText(message);
-            formatted.Pop();
-            Contents.AddMessage(formatted);
+            if (_currentDrag == DragMode.None)
+            {
+                var cursor = CursorShape.Arrow;
+                var previewDragMode = GetDragModeFor(args.RelativePosition);
+                switch (previewDragMode)
+                {
+                    case DragMode.Bottom:
+                        cursor = CursorShape.VResize;
+                        break;
+
+                    case DragMode.Left:
+                        cursor = CursorShape.HResize;
+                        break;
+
+                    case DragMode.Bottom | DragMode.Left:
+                        cursor = CursorShape.Crosshair;
+                        break;
+                }
+
+                DefaultCursorShape = cursor;
+            }
+            else
+            {
+                var top = Rect.Top;
+                var bottom = Rect.Bottom;
+                var left = Rect.Left;
+                var right = Rect.Right;
+                var (minSizeX, minSizeY) = CombinedMinimumSize;
+                if ((_currentDrag & DragMode.Bottom) == DragMode.Bottom)
+                {
+                    bottom = Math.Max(args.GlobalPosition.Y + _dragOffsetBottomRight.Y, top + minSizeY);
+                }
+
+                if ((_currentDrag & DragMode.Left) == DragMode.Left)
+                {
+                    var maxX = right - minSizeX;
+                    left = Math.Min(args.GlobalPosition.X - _dragOffsetTopLeft.X, maxX);
+                }
+
+                ClampSize(left, bottom);
+            }
+        }
+
+        protected override void UIScaleChanged()
+        {
+            base.UIScaleChanged();
+            ClampAfterDelay();
+        }
+
+        private void ClydeOnOnWindowResized(WindowResizedEventArgs obj)
+        {
+            ClampAfterDelay();
+        }
+
+        private void ClampAfterDelay()
+        {
+            if (!_lobbyMode)
+                _clampIn = 2;
+        }
+
+        protected override void FrameUpdate(FrameEventArgs args)
+        {
+            base.FrameUpdate(args);
+            if (_lobbyMode) return;
+            // we do the clamping after a delay (after UI scale / window resize)
+            // because we need to wait for our parent container to properly resize
+            // first, so we can calculate where we should go. If we do it right away,
+            // we won't have the correct values from the parent to know how to adjust our margins.
+            if (_clampIn <= 0) return;
+            _clampIn -= 1;
+            if (_clampIn == 0) ClampSize();
+        }
+
+        private void ClampSize(float? desiredLeft = null, float? desiredBottom = null)
```

Comment:
Is Control.ApplySizeConstraints able to replace this or maybe even remove this completely if Control automatically handles clamping now?

Reply:
I don't think so?